### ISSUE TITLE
Moving OCI integration from experimental to full feature

### ISSUE THREAD

**Issue by scottrigby** [helm/helm#10393]
Most of the work outlined in [hip-0006](https://github.com/helm/community/blob/main/hips/hip-0006.md) has been finished.

The upstream [ORAS Go](https://github.com/oras-project/oras-go) project that Helm imports now has a stable `1.0.0` release, with a backward compatibility guarantee. This was the last major step in unblocking releasing Helm OCI integration as a full feature.

## Remaining steps

- [x] Documentation only: https://github.com/helm/helm/issues/10169#issuecomment-976973309
    - #1249
- [x] #10534 
    - #10535
- [x] #10166
    - https://github.com/helm/helm/pull/10521
- [x] #10156
    - https://github.com/oras-project/oras-go/pull/92
    - #10536 
- [x] #9694
    - https://github.com/oras-project/oras-go/pull/89
    - https://github.com/helm/helm/pull/10527
- [x] #9188
    - #10546 to be merged last, after all other work above is complete

## Not identified as blockers

However, if someone has time to test these oci-related PRs, they may still get into 3.8.0.

- [ ] #6324
    - PR ready for review 👀: #10408 
- [ ] #10175 
    - PR ready for review 👀  #10430

## Follow-up

This issue is only concerned with comprehensively listing any blockers for moving OCI support to a full feature. New features can be added after it is moved from experimental. After that point, only backward-compatible changes will be accepted until Helm 4.

- Join the discussion about backward-compatible features here: https://github.com/helm/helm/issues/10312

_Update (by mattfarina) 12/3: Connecting with technosophos and jdolitsky we reviewed the open issues from bacongobbler and added two additional tasks to the remaining steps._

**Comment by bacongobbler** [comment]
> Anything else?

Here's my laundry list... I started by searching the issue queue. Here's the search terms I used to look for open tickets:

https://github.com/helm/helm/issues?q=is%3Aissue+is%3Aopen+oci
https://github.com/helm/helm/issues?page=1&q=is%3Aissue+is%3Aopen+docker
https://github.com/helm/helm/issues?q=is%3Aissue+is%3Aopen+registry

This was just a casual search... It might be worth triaging the entire queue just to make sure no stone is left unturned.

Anyways, I saw a few tickets that might be worth evaluating (some have been discussed at length during previous developer calls):

- https://github.com/helm/helm/issues/10013
- https://github.com/helm/helm/issues/9694
- https://github.com/helm/helm/issues/6324
- https://github.com/helm/helm/issues/6214
- https://github.com/helm/helm/issues/10156
- https://github.com/helm/helm/issues/10123
- https://github.com/helm/helm/issues/10166

#9694 in particular seems like a critical code path that should be addressed. No version range support for both `helm install` and dependencies is a big regression compared to chart repositories because users must now know what version they need to install. `helm install wordpress --generate-name` is currently not possible with OCI. Users MUST declare what version to install. To me that's a huge regression.

I know we've discussed `helm search` at great length during a few developer calls, but we need an answer whether the maintainers are going to work on a solution, or state that `helm search` does not support OCI registiries (and mark that as a ❌when comparing to chart repositories). Choosing to implement it could affect how information is stored in the registry, which would constitute a breaking change.

#10166 might also be worth taking a closer look as it pertains to how chart versions are stored in the registry.

I'd also recommend a bake-off between what you can do with chart repositories vs OCI registries before we can mark it as graduated. If only to inform users of the benefits and tradeoffs they make when choosing one or the other.

https://github.com/helm/helm/blob/main/CONTRIBUTING.md#semantic-versioning states that "Chart repository functionality MUST be backward compatible". But it doesn't state anything about OCI registry functionality.

Finally... The Helm documentation on chart registries would also be worth taking a closer look to make sure everything is up to date. Things like "why someone would choose OCI registries over chart repositories" would be worth documenting.

**Comment by bacongobbler** [comment]
Doc suggestion: "migrating charts from a chart repository to OCI". Either as a plugin similar to https://github.com/helm/helm-2to3 or just a guide on what commands to run.

**Comment by bacongobbler** [comment]
I'd also recommend looking at the oras issue queue to search for any open tickets that may be worth investing time fixing upstream. This one seems related to #9694: https://github.com/oras-project/oras-go/issues/19

**Comment by pytimer** [comment]
Implement helm push/pull support insecure OCI registry in https://github.com/helm/helm/pull/10408, fix  [How can I push chart to an insecure OCI registry with helm v3 #6324](https://github.com/helm/helm/issues/6324)

**Comment by scottrigby** [comment]
Updates:

Folks are working hard on these issues leading up to and potentially during the holidays. In order for this to be moved from experimental to full feature in the next minor release (3.8.0) all blockers would have to be complete by the scheduled RC date (Jan 10, 2022). If not, we can re-target to the next minor release (3.9.0). Let's see what we can accomplish!

I have removed "Forthcoming issue: @jdolitsky re DOCKER_CONFIG" from our checklist in the issue description, and added a comment here with the current plan: https://github.com/helm/helm/issues/10156#issuecomment-999821534 @jdolitsky you can cross this off your list to explain in an issue ✨ 

@pytimer I have added #6324 and your PR #10408 to a new category above called "Not yet identified as a blocker", since I'm not yet clear whether this should be a blocker for moving OCI from experimental to full feature, or a nice to have by then and alternatively could be a follow-up. Will see if we can get this reviewed.

**Comment by pytimer** [comment]
@scottrigby Thanks, i know. :-)  If something should i do, please tell me.

**Comment by KlavsKlavsen** [comment]
I assume target of 3.8.0 was not reached with this.. So :crossed_fingers: for 3.9.0 :)

**Comment by scottrigby** [comment]
@KlavsKlavsen today's RC was pushed back two days to accommodate completing this. This still gives a full week between RC and 3.8.0 to fix regressions. It is still possible we won't make the deadline, but a number of us are putting in a lot of work to make this happen, and it's looking very good right now.

**Comment by scottrigby** [comment]
## Update: we are nearly there!

- https://github.com/helm/helm/pull/10527 PR is solid. New test added. All checks pass. Just awaiting one other Helm maintainer to review/approve then I will merge
- Oras maintainers can then cut `oras-go` `v1.1.0`, since we now confirmed tag API, auth, and docker config fallback all work now in `v1.1.0-rc3`
- Last step will be to finish moving the package out of experimental, remove the experimental env var, and feature flag gates. Thanks to @runyontr for starting this – @dperaza4dustbit is planning to bring in his commits into a separate PR and finish that tomorrow. If for some reason David is pulled away, I can do that.

## How everyone else here can help
Assuming we will get this into the Helm RC, we will want everyone interested in this feature to test it. There will be a week to fix regressions found before 3.8.0. Will confirm in this issue if/when that is indeed the case.

We are looking good! 🤞 will keep issue updated

**Comment by danquah** [comment]
Awesome! Just say the word, I'll be happy to throw some hours into testing out a project we have that currently uses the experimental flag to push some Helm Charts to both Github Packages and Azure Container Registry

**Comment by KlavsKlavsen** [comment]
link to Helm RC - and we'll update our docker images and CI pipelines to use this for OCI and see how it works

**Comment by scottrigby** [comment]
This is complete! Closing this meta tracking issue, and will add a comment once RC is out. Thank you all for your efforts, and thanks in advance for testing the oci feature in the RC ✨

**Comment by scottrigby** [comment]
📣  [v3.8.0-rc.1](https://github.com/helm/helm/releases/tag/v3.8.0-rc.1) is out! Please test and promote widely for the most oci-related user regression testing possible before the [3.8.0 milestone](https://github.com/helm/helm/milestone/113) on 1/19 .
